### ISSUE TITLE
PureComponent's shallowEqual

### ISSUE THREAD

**Issue by RayChristian** [facebook/react#13468]
```
class App extends React.PureComponent {
  state={
    ele: {a:'1'}
  }
  trigger=()=>{
    console.log('click')
    this.setState({ele: {a:'1'}})
  }
  render() {
    console.log('render')
    return (
      <div onClick={this.trigger}>click here to trigger setState</div>
    );
  }
}
```
A simple demo shows that click event triggers render every time, just like triple-equals.
[demo](https://codepen.io/ruithm/pen/PdqRNP?editors=1111)

However, according to [source code](https://github.com/facebook/fbjs/blob/master/packages/fbjs/src/core/shallowEqual.js )React.PureComponent’s shouldComponentUpdate() shallowly compares the objects like following.

>  Performs equality by iterating through keys on an object and returning false when any key has values which are not strictly equal between the arguments. Returns true when the values of all keys are strictly equal.

**Comment by NE-SmallTown** [comment]
> PureComponent shallow equal is related to props not state,

This is not correct, the [doc](https://reactjs.org/docs/react-api.html#reactpurecomponent) says " it with a shallow prop and **state** comparison."

Or just from the src code(https://github.com/facebook/react/blob/0beb2ee76b96ffb1c15a1770a6907ea1cd5a5696/packages/react-reconciler/src/ReactFiberClassComponent.js#L261)  is same

**Comment by jquense** [comment]
Oops, yes your right :P not thinking clearly. The second point still holds, it's not shallow ly equal in your example

**Comment by RayChristian** [comment]
```
/**
 * Performs equality by iterating through keys on an object and returning false
 * when any key has values which are not strictly equal between the arguments.
 * Returns true when the values of all keys are strictly equal.
 */
function shallowEqual(objA: mixed, objB: mixed): boolean {
  if (is(objA, objB)) {
    return true;
  }

  if (typeof objA !== 'object' || objA === null ||
      typeof objB !== 'object' || objB === null) {
    return false;
  }

  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);

  if (keysA.length !== keysB.length) {
    return false;
  }

  // Test for A's keys different from B.
  for (let i = 0; i < keysA.length; i++) {
    if (
      !hasOwnProperty.call(objB, keysA[i]) ||
      !is(objA[keysA[i]], objB[keysA[i]])
    ) {
      return false;
    }
  }

  return true;
}
```
PureComponent shallow equal is related to state, have wrote a demo to prove. Also, if shallowEqual is just like '===', what is meaning of the source code above?

**Comment by gaearon** [comment]
`this.state` is not shallowly equal to `prevState` because `this.state.ele` and `prevState.ele` are two different objects. Therefore the shallow equality check fails.

**Comment by RayChristian** [comment]
@gaearon Actually I just want to know is there any difference between PureComponent's shallowEqual and '===', because according to the [src code](https://github.com/facebook/fbjs/blob/master/packages/fbjs/src/core/shallowEqual.js) there is.

**Comment by gaearon** [comment]
If you mean individual properties, afaik they're compared using `Object.is` algorithm (which is slightly different from `===` because it also handles NaN).
